### ISSUE TITLE
XCF signaling stopped working.

### ISSUE THREAD

**Issue by edtgek** [Fish-Git/hyperion#15]
Hi.
With Hercules version ga36e0bbc ([8873](http://www.softdevlabs.com/hyperion#8873), dated 2017-07-08) I reported an issue with “XCF signaling stopped working” (Issue #11).

With version g6359717a ([8904](http://www.softdevlabs.com/hyperion#8904), dated 2017-11-04) it seems that same issue is back. I have the same problem.

With version  gb6106b1  (which was a private build at commit b6106b102b740755a9706ba6fc38df1b33e30463 dated [Jul 29, 2017](https://github.com/Fish-Git/hyperion/commits/master) that includes the [fix for Issue 11](https://github.com/Fish-Git/hyperion/commit/71d9a74804a79d544b7b67d37093177c07b09255)), everything with XCF was working fine for me.

Hercules configuration file for system E001:
```
    0E24 CTCE 31880 localhost 31882 
    0E25 CTCE 31884 localhost 31886
```

Hercules configuration file for system E002:
```
    0E24 CTCE 31882 localhost 31880
    0E25 CTCE 31886 localhost 31884
```

![image](https://user-images.githubusercontent.com/30401203/32442953-8dd6970a-c2fd-11e7-80ad-6a9df8c0014f.png)

Goran Eklof
mail: goran@ekloef.se

**Comment by Fish-Git** [comment]
Hi Göran!

It appears Peter Jansen's latest changes to the CTCE driver that I [merged](https://github.com/Fish-Git/hyperion/pull/13) may have reverted my earlier fix.

I need to do some research to see if maybe I can figure out how it broke and how to fix it (which may require me to contact Peter for help).

In the mean time, could you please do us a favor to help us please?

I would like to see a CCW trace on devices E24 and E25 for _both_ commit 9269d74bf5dd81710e4216125f685219106533c0 _(without_ Peter's changes) and commit 6359717ac516237d4c2f61e70273b69ada6e6898 (_with_ Peter's changes).

Make sure you do a CCW trace (`t+e24`, `t+e25`) when you run your test on _each version_. The goal here is to compare how each behaves, both with and without Peter's changes, so I need to see the CCW trace for _both_ test runs (both _with_ and _without_ Peter's changes).

Thanks!

**Comment by Peter-J-Jansen** [comment]
Hi Fish and Göran,

 

I think I may have found and fixed the “XCF signaling stopped working” problem.

 

First I’d like to thank Fish for merging my previous CTCE fixes into his repository, which I’ve been referring to as “SDL-hyperion” to avoid confusion with the other Hyperion repository. When I browsed the resulting “ctcadpt.c” code, at first I thought it would be OK, but then I realized that the extra logic Fish introduced around “ctce_haltorclear” in the CTCE_Send routine would cause problems on the “send” side. Hence that I regressed that, and tested the result. XCF (and z/VM’s ISFC too) are now working again. The fix is available here: https://github.com/Peter-J-Jansen/SDL-hyperion, and I have already entered a new “pull request” so that Fish may merge it with his SDL-hyperion repository.

 

Any comments are welcome of course.

 

Thanks,

 

Peter

**Comment by Peter-J-Jansen** [comment]
Hi Fish and Göran,

 

I think I may have found and fixed the “XCF signaling stopped working” problem.

 

First I’d like to thank Fish for merging my previous CTCE fixes into his repository, which I’ve been referring to as “SDL-hyperion” to avoid confusion with the other Hyperion repository. When I browsed the resulting “ctcadpt.c” code, at first I thought it would be OK, but then I realized that the extra logic aroung “ctce_haltorclear” in the CTCE_Send routine would cause problems on the “send” side. Hence that I regressed that, and tested the result. XCF (and z/VM’s ISFC) are now working again. The fix is available here : https://github.com/Peter-J-Jansen/SDL-hyperion , and I have already entered a new “pull request” so that Fish may merge it with his SDL-hyperion repository.

 

Any comments are welcome of course.

 

Thanks,

 

Peter

**Comment by Peter-J-Jansen** [comment]
Oops, sorry for sending the same e-mail twice.

**Comment by Fish-Git** [comment]
> Oops, sorry for sending the same e-mail twice.

No problem.  As the repository owner, I can delete any comment I want.

Your duplicate comment has been deleted!  Problem solved! &nbsp; `:)`

___
_**FYI:** I would personally prefer that everyone _**not**_ use email when responding/replying to an issue, and instead use GitHub's web based user interface._

Yes, I realize it may be faster and easier to respond via email and slightly more bothersome to use GitHub's web interface instead, but here's the problem: GitHub does not allow me to edit email based replies using [markdown (styling)](https://guides.github.com/features/mastering-markdown/).  It only allow markdown for comments entered via their _web based user interface_.

Forgive me, but oftentimes the comments added by others are not well formatted, and I prefer to be able to edit them so they are easier to read and understand (adding italics, inline code, etc) .  I cannot do that when you reply via email.

You can continue to reply via email if you want.  That's okay.   _I'm not asking or demanding that you to change your behavior!_

I'm simply letting you know about my own personal preference, that's all.

Whether or not you feel my personal preference is important enough to you to change your behavior is entirely up to you. &nbsp; `;-)`

**Comment by Fish-Git** [comment]
> ... but then I realized that the extra logic Fish introduced around “ctce_haltorclear” in the CTCE_Send routine would cause problems on the “send” side. [...]

That was going to be my second guess! &nbsp; `;-)`

I wasn't sure whether it was my [attempt to the fix the HaltOrClear logic in the various communication device handlers](https://github.com/Fish-Git/hyperion/commit/5eb60c6b762a16ad2c77a9b5d610b0024861c68a) (LCS, CTCI, CTCE and PTP) or whether it was something _you_ did in your original merge request (which was based on a version of my SDL Hyperion _without_ my "fix").  I was hoping Göran could do some tests for us to determine which it was.  (I am unable to test CTCE myself.)

I only suspected your code since my code is always perfect and 100% bug free.  Only other people make mistakes.  I never do.  &nbsp;`;-)`
&lt;/sarcasm&gt;

Anyway, _**THANK YOU Peter**_, for finding and fixing this bug so quickly!

And please accept my apology for temporarily breaking your truly awesome code (and it _**is**_ truly _awesome!!!_).

I will merge your fix right away, and close this issue once I do.

Thanks again for everything, Peter.

You are a good friend. &nbsp; ``:`)``

**Comment by Fish-Git** [comment]
Resolved by merge "Fixed "XCF signalling stopped working" #16 (commit 1b91101977d28a05e87df1a3243b4756bf77edba).

Closing this issue as resolved.

Thanks again, Peter!

I will build and upload a new Windows pre-built with this fix in it as soon as I can.
